src/service/layers-impl.js

```diff
@@ -0,0 +1,634 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {layoutRectLtwh} from '../layout-rect';
+import {dev} from '../log';
+import {filterSplice} from '../utils/array';
+import {Services} from '../services';
+import {registerServiceBuilderForDoc} from '../service';
+import {computedStyle} from '../style';
+
+const LAYOUT_PROP = '__AMP_LAYOUT';
+
+/**
+ * The Size of an element.
+ *
+ * @typedef {{
+ *   height: number,
+ *   width: number,
+ * }}
+ */
+export let SizeDef;
+
+/**
+ * The offset Position of an element.
+ *
+ * @typedef {{
+ *   left: number,
+ *   top: number,
+ * }}
+ */
+export let PositionDef;
+
+/**
+ * Creates a Size.
+ *
+ * @param {number} width
+ * @param {number} height
+ * @return {!SizeDef}
+ */
+function SizeWh(width, height) {
```

Comment:
Good fit for [`@struct`](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#struct).